streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -883,12 +953,26 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
                                                    final ValueJoiner<? super V1, ? super V2, ? extends R> joiner,
                                                    final JoinWindows windows,
                                                    final Joined<K1, V1, V2> joined) {
-            final String thisWindowStreamName = builder.newProcessorName(WINDOWED_NAME);
-            final String otherWindowStreamName = builder.newProcessorName(WINDOWED_NAME);
-            final String joinThisName = rightOuter ? builder.newProcessorName(OUTERTHIS_NAME) : builder.newProcessorName(JOINTHIS_NAME);
-            final String joinOtherName = leftOuter ? builder.newProcessorName(OUTEROTHER_NAME) : builder.newProcessorName(JOINOTHER_NAME);
-            final String joinMergeName = builder.newProcessorName(MERGE_NAME);
-
+            final NamedInternal namedInternal = new NamedInternal(joiner);
+
+            final String thisWindowStreamName =  namedInternal.suffixWithOrElseGet(
```

Comment:
I'd suggest we use the suffix of those nodes as `-this/other-windowed` and `-this/other-join` and `-merge` for simplicity.